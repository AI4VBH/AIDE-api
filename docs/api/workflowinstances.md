# Workflow Instances

## `PUT /workflwoinstances/:workflow_instance_id/executions/:execution_id/acknowledge`

Acknowledge task errors for a given execution of a workflow.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:workflow_instance_id` | The workflow instance ID |
| URL | `:execution_id` | The task execution ID |

### Response

| Code | Description |
|------|-------------|
| `200` | Returns acknowledged `MonaiWorkflowInstance` |
| `404` | Workflow instance not found or task execution not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request PUT 'http://localhost:5000/workflwoinstances/3ef05fc1-dd5e-402a-b382-a557de1e383c/executions/04533a0b-d43d-47d4-8aac-1efaca39b778/acknowledge'
```

### Example Response

```json
{
  "acknowledged_workflow_errors": "",
  "ae_title": "MONAI",
  "bucket_id": "monaideploy",
  "id": "a67a7af7-068b-44b8-a81b-def7b3e5403b",
  "input_metadata": {},
  "payload_id": "d801bd68-381d-4ff3-95a8-85c46380e075",
  "start_time": "2022-09-22T12:45:13.102Z",
  "status": "Failed",
  "tasks": [
    {
      "acknowledged_task_errors": "2022-10-03T12:00:00.001Z",
      "execution_id": "3b9d94b9-4285-45d4-bea9-491fa62b8f91",
      "execution_stats": {},
      "input_artifacts": {
        "input-dicom": "d801bd68-381d-4ff3-95a8-85c46380e075/dcm",
      },
      "input_parameters": {},
      "next_timeout": "2022-09-22T12:45:13.102Z",
      "output_artifacts": {},
      "output_directory": "d801bd68-381d-4ff3-95a8-85c46380e075/workflows/a67a7af7-068b-44b8-a81b-def7b3e5403b/3b9d94b9-4285-45d4-bea9-491fa62b8f91",
      "previous_task_id": "",
      "reason": "Connection time-out",
      "result": {},
      "status": "failed",
      "task_id": "mean-pixel-calc",
      "task_plugin_arguments": {},
      "task_start_time": "2022-09-22T12:45:13.102Z",
      "task_type": "argo",
      "timeout_interval": 0,
      "workflow_instance_id": "a67a7af7-068b-44b8-a81b-def7b3e5403b"
    }
  ],
  "workflow_id": "0c262733-05a1-4811-9d9a-b5cadb5a475a"
}
```