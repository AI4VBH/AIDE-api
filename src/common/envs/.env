PORT=5000

KEYCLOAK_URL='http://0.0.0.0:8443'
KEYCLOAK_MAIN_REALM='master'
KEYCLOAK_REALM='Demo-Realm'
KEYCLOAK_CLIENTID='nest-app'
KEYCLOAK_SECRET='83790b4f-48cd-4b6c-ac60-451a918be4b9'
KEYCLOAK_STATIC_ROLES='["admin", "clinician"]'

DATABASE_HOST='localhost'
DATABASE_NAME='nest_api'
DATABASE_USER='postgres'
DATABASE_PASSWORD='admin'
DATABASE_PORT=5432

ELASTIC_HOST=''
ELASTIC_PORT=''
ELASTIC_USE_SSL=''
ELASTIC_USERNAME=''
ELASTIC_PASSWORD=''
ELASTIC_INDEX=''

MONAI_API_HOST='https://localhost:8081/'
MIG_API_HOST='http://localhost:5010/'

MINIO_HOST='localhost'
MINIO_PORT=9001
MINIO_USE_SSL=false
MINIO_BUCKET='monaideploy'
MINIO_ACCESS_KEY='wOzmbchico52AMp3'
MINIO_SECRET_KEY='zyoT9SiMLiTCvNT0DS636FMiN4yi2258'
