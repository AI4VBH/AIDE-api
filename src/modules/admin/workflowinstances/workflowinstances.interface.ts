/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

export interface MonaiWorkflowInstance {
  id: string;
  ae_title: string;
  workflow_id: string;
  workflow_name: string;
  start_time: string;
  payload_id: string;
  status: string;
  bucket_id: string;
  acknowledged_workflow_errors: string;
  tasks: MonaiWorkflowTask[];
  input_metadata: [{ [name: string]: string }];
}

export interface MonaiWorkflowTask {
  execution_id: string;
  workflow_instance_id: string;
  previous_task_id: string;
  task_id: string;
  task_type: string;
  task_start_time: string;
  status: string;
  reason: string;
  acknowledged_task_errors: string;
  input_artifacts: [{ [name: string]: string }];
  output_artifacts: [{ [name: string]: string }];
  output_directory: string;
  result: [{ [name: string]: string }];
  next_timeout: string;
  timeout_interval: number;
  execution_stats?: [{ [name: string]: string }];
}
