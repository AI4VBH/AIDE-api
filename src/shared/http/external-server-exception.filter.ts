import {
  ArgumentsHost,
  Catch,
  ExceptionFilter,
  HttpException,
  HttpStatus,
  Logger,
} from '@nestjs/common';
import { Response } from 'express';
import { AxiosError } from 'axios';
import { WorkflowServiceException } from 'modules/workflows/workflow.service.exceptions';
import { MinoiClientException } from 'shared/minio/minio-client';
import {
  ExecutionsServiceException,
  ExecutionsServiceExceptionType,
} from 'modules/admin/executions/executions.service.exceptions';

@Catch(Error)
export default class ExternalServerExceptionFilter implements ExceptionFilter {
  catch(exception: Error, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<Response>();
    const logger = new Logger('ExceptionFilter');

    logger.error(exception, JSON.stringify(exception, null, 2));

    if (exception instanceof HttpException) {
      const status = exception.getStatus();

      return response.status(status).json({
        statusCode: status,
        message: exception.message,
      });
    }

    if (Object.keys(exception).includes('config')) {
      const { message, status } = (
        exception as AxiosError
      ).toJSON() as ResponseException;

      if (
        status === HttpStatus.REQUEST_TIMEOUT ||
        status >= HttpStatus.INTERNAL_SERVER_ERROR
      ) {
        return response.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
          statusCode: HttpStatus.INTERNAL_SERVER_ERROR,
          message: 'An issue occurred with the MONAI service',
        });
      }

      if (
        status === HttpStatus.BAD_REQUEST ||
        status === HttpStatus.NOT_FOUND
      ) {
        return response
          .status(status)
          .json((exception as AxiosError).response.data);
      }

      return response.status(status ?? HttpStatus.INTERNAL_SERVER_ERROR).json({
        statusCode: status ?? HttpStatus.INTERNAL_SERVER_ERROR,
        message,
      });
    }

    if (exception instanceof WorkflowServiceException) {
      return response.status(HttpStatus.BAD_REQUEST).json({
        statusCode: HttpStatus.BAD_REQUEST,
        message: exception.message,
      });
    }

    if (exception instanceof MinoiClientException) {
      return response.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
        statusCode: HttpStatus.INTERNAL_SERVER_ERROR,
        message: 'An issue occurred with the MINIO service',
      });
    }

    if (exception instanceof ExecutionsServiceException) {
      if (exception.type === ExecutionsServiceExceptionType.MISSING_TASK) {
        return response.status(HttpStatus.NOT_FOUND).json({
          statusCode: HttpStatus.NOT_FOUND,
          message: exception.message,
        });
      }
    }

    response.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
      statusCode: 500,
      message: 'Internal Server Error',
    });
  }
}

type ResponseException = {
  status: number;
  message: string;
};
