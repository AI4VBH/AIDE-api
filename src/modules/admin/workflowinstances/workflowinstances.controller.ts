import { Controller, Put, Inject, Param, UseFilters } from '@nestjs/common';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { WorkflowInstancesService } from './workflowinstances.service';

@Controller('workflowinstances')
@UseFilters(ExternalServerExceptionFilter)
export class WorkflowInstanceController {
  @Inject(WorkflowInstancesService)
  private readonly wfiService: WorkflowInstancesService;

  @Put(':workflow_instance_id/executions/:execution_id/acknowledge')
  acknowledgeTaskError(
    @Param('workflow_instance_id') workflow_instance_id,
    @Param('execution_id') execution_id,
  ) {
    return this.wfiService.acknowledgeTaskError(
      workflow_instance_id,
      execution_id,
    );
  }
}
