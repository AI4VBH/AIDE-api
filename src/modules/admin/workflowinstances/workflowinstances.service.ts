import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { AxiosError, AxiosResponse } from 'axios';
import { lastValueFrom } from 'rxjs';
import { MonaiWorkflowInstance } from './workflowinstances.interface';

@Injectable()
export class WorkflowInstancesService {
  @Inject(HttpService)
  private readonly httpService: HttpService;
  private readonly CONTROLLER: string = 'workflowinstances';
  async acknowledgeTaskError(
    workflow_instance_id: string,
    execution_id: string,
  ): Promise<MonaiWorkflowInstance> {
    const response = await lastValueFrom(
      this.httpService.put<MonaiWorkflowInstance>(
        `${this.CONTROLLER}/${workflow_instance_id}/executions/${execution_id}/acknowledge`,
      ),
    );

    return response.data;
  }

  async getAcknowledgedTaskErrors(
    sinceDate: string,
  ): Promise<MonaiWorkflowInstance[]> {
    const response = await lastValueFrom(
      this.httpService.get<MonaiWorkflowInstance[]>(
        `workflowinstances/failed?acknowledged=${sinceDate}`,
      ),
    );

    return response.data;
  }
}
