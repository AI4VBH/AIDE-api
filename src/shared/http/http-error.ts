import { HttpStatus } from "@nestjs/common/enums";

export class HttpError extends Error {
    constructor(statusCode: HttpStatus, message: string) {
      super(message);
  
      this.statusCode = statusCode;
    }
  
  readonly statusCode: HttpStatus;
  }
  
  