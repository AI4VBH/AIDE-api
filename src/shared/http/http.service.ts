/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Injectable, Inject } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { HttpModuleOptionsFactory, HttpModuleOptions } from '@nestjs/axios';

@Injectable()
export class HttpConfigService implements HttpModuleOptionsFactory {
  @Inject(ConfigService)
  private readonly config: ConfigService;

  createHttpOptions(): HttpModuleOptions {
    let encodeToken = Buffer.from(`${this.config.get<string>('MONAI_API_USER')}:${this.config.get<string>('MONAI_API_KEY')}`)
      .toString('base64');
    return {
      timeout: 5000,
      maxRedirects: 5,
      headers: {'Authorization': `Basic ${encodeToken}`},
      baseURL: this.config.get<string>('MONAI_API_HOST'),
    };
  }
}
