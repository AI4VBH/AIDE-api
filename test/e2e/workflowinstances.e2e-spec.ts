/*
 * Copyright 2022 Guyâ€™s and St Thomasâ€™ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { HttpModule } from '@nestjs/axios';
import { ConfigModule } from '@nestjs/config';
import * as request from 'supertest';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { HttpConfigService } from 'shared/http/http.service';
import WorkflowInstanceMocks from '../test_data/mocks/workflow-instances/workflow-instances';
import { WorkflowInstanceController } from 'modules/admin/workflowinstances/workflowinstances.controller';
import { WorkflowInstancesService } from 'modules/admin/workflowinstances/workflowinstances.service';

const server = setupServer();
const testMonaiBasePath = 'https://localhost:7337';

describe('/workflowinstances integration Tests', () => {
  let app: INestApplication;

  beforeAll(() => {
    server.listen({
      onUnhandledRequest: 'bypass',
    });
    server.printHandlers();
  });

  afterAll(() => {
    server.close();
  });

  beforeEach(async () => {
    server.resetHandlers();

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
          // we mock the path to the MONAI API
          load: [
            () => ({
              MONAI_API_HOST: testMonaiBasePath,
            }),
          ],
        }),
        // we register the HttpService with no stubbing! ðŸŽ‰
        HttpModule.registerAsync({
          useClass: HttpConfigService,
        }),
      ],
      controllers: [WorkflowInstanceController],
      providers: [WorkflowInstancesService],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it('(PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge with returned data', async () => {
    server.use(
      rest.put(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge`,
        (_request, response, context) => {
          return response(
            context.json(WorkflowInstanceMocks.workflowInstance2),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).put(
      '/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(200);
  });

  it('(PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Non existent workflow instance id', async () => {
    server.use(
      rest.put(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge`,
        (_request, response, context) => {
          return response(
            context.status(404),
            context.json(WorkflowInstanceMocks.nonExistentWorkflowInstanceId),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).put(
      '/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(404);
  });

  it('(PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Non existent execution id', async () => {
    server.use(
      rest.put(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/3b9d94b9-4285-45d4-bea9-491fa62b8g66/acknowledge`,
        (_request, response, context) => {
          return response(
            context.status(404),
            context.json(WorkflowInstanceMocks.nonExistentExecutionId),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).put(
      '/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/3b9d94b9-4285-45d4-bea9-491fa62b8g66/acknowledge',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(404);
  });

  it('(PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Invalid workflow instance ID', async () => {
    server.use(
      rest.put(
        `${testMonaiBasePath}/workflowinstances/invalidGUID/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge`,
        (_request, response, context) => {
          return response(
            context.status(400),
            context.json(WorkflowInstanceMocks.invalidWorkflowInstanceId),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).put(
      '/workflowinstances/invalidGUID/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(400);
  });

  it('(PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Invalid execution ID', async () => {
    server.use(
      rest.put(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/invalidGUID/acknowledge`,
        (_request, response, context) => {
          return response(
            context.status(400),
            context.json(WorkflowInstanceMocks.invalidExecutionId),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).put(
      '/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/invalidGUID/acknowledge',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - correct status when MONAI gives general error with code %s',
    async (code) => {
      server.use(
        rest.put(
          `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).put(
        '/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge',
      );
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );
});
