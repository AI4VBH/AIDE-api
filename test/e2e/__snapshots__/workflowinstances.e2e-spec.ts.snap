// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Generic MONAI errors 1`] = `
{
  "message": "An issue occurred with the MONAI service",
  "statusCode": 500,
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Generic MONAI errors 2`] = `
{
  "message": "An issue occurred with the MONAI service",
  "statusCode": 500,
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Generic MONAI errors 3`] = `
{
  "message": "An issue occurred with the MONAI service",
  "statusCode": 500,
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Generic MONAI errors 4`] = `
{
  "message": "An issue occurred with the MONAI service",
  "statusCode": 500,
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Generic MONAI errors 5`] = `
{
  "message": "An issue occurred with the MONAI service",
  "statusCode": 500,
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Generic MONAI errors 6`] = `
{
  "message": "An issue occurred with the MONAI service",
  "statusCode": 500,
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Invalid execution ID 1`] = `
{
  "detail": "Failed to validate execution ID, not a valid GUID",
  "instance": "/workflowinstances/invalidGUID/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-7d36d37cd21021e68d552618bc5bfc3c-335958bf8d6d9ad0-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Invalid workflow instance ID 1`] = `
{
  "detail": "Failed to validate workflow instance ID, not a valid GUID",
  "instance": "/workflowinstances/invalidGUID/executions/3b9d94b9-4285-45d4-bea9-491fa62b8f91/acknowledge",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-7d36d37cd21021e68d552618bc5bfc3c-335958bf8d6d9ad0-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Non existent execution id 1`] = `
{
  "detail": "Failed to validate id, execution not found",
  "instance": "/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b/executions/3b9d94b9-4285-45d4-bea9-491fa62b8g66/acknowledge",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge - Non existent workflow instance id 1`] = `
{
  "detail": "Failed to validate id, workflow instance not found",
  "instance": "/workflows/a67a7af7-068b-44b8-a81b-def7b3e5403b",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/workflowinstances integration Tests (PUT) /workflowinstances/{workflowInstanceId}/executions/{executionId}/acknowledge with returned data 1`] = `
{
  "acknowledged_workflow_errors": "",
  "ae_title": "MONAI",
  "bucket_id": "monaideploy",
  "id": "a67a7af7-068b-44b8-a81b-def7b3e5403b",
  "input_metadata": {},
  "payload_id": "d801bd68-381d-4ff3-95a8-85c46380e075",
  "start_time": "2022-09-22T12:45:13.102Z",
  "status": "Failed",
  "tasks": [
    {
      "acknowledged_task_errors": "2022-10-03T12:00:00.001Z",
      "execution_id": "3b9d94b9-4285-45d4-bea9-491fa62b8f91",
      "execution_stats": {},
      "input_artifacts": {
        "input-dicom": "d801bd68-381d-4ff3-95a8-85c46380e075/dcm",
      },
      "input_parameters": {},
      "next_timeout": "2022-09-22T12:45:13.102Z",
      "output_artifacts": {},
      "output_directory": "d801bd68-381d-4ff3-95a8-85c46380e075/workflows/a67a7af7-068b-44b8-a81b-def7b3e5403b/3b9d94b9-4285-45d4-bea9-491fa62b8f91",
      "previous_task_id": "",
      "reason": "Connection time-out",
      "result": {},
      "status": "failed",
      "task_id": "mean-pixel-calc",
      "task_plugin_arguments": {},
      "task_start_time": "2022-09-22T12:45:13.102Z",
      "task_type": "argo",
      "timeout_interval": 0,
      "workflow_instance_id": "a67a7af7-068b-44b8-a81b-def7b3e5403b",
    },
    {
      "acknowledged_task_errors": "",
      "execution_id": "47d73804-ea9b-4aaf-b375-bd19ae8feaae",
      "execution_stats": {},
      "input_artifacts": {
        "report-dicom": "d801bd68-381d-4ff3-95a8-85c46380e075/workflows/a67a7af7-068b-44b8-a81b-def7b3e5403b/3b9d94b9-4285-45d4-bea9-491fa62b8f91/report-dicom",
      },
      "input_parameters": {},
      "next_timeout": "2022-09-22T12:45:59.689Z",
      "output_artifacts": {},
      "output_directory": "d801bd68-381d-4ff3-95a8-85c46380e075/workflows/a67a7af7-068b-44b8-a81b-def7b3e5403b/47d73804-ea9b-4aaf-b375-bd19ae8feaae",
      "previous_task_id": "mean-pixel-calc",
      "reason": "Connection time-out",
      "result": {},
      "status": "Failed",
      "task_id": "export-report",
      "task_plugin_arguments": {},
      "task_start_time": "2022-09-22T12:45:59.689Z",
      "task_type": "export",
      "timeout_interval": 0,
      "workflow_instance_id": "a67a7af7-068b-44b8-a81b-def7b3e5403b",
    },
  ],
  "workflow_id": "0c262733-05a1-4811-9d9a-b5cadb5a475a",
}
`;
